src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -452,7 +466,24 @@ public async Task ReadResponseHeadersAsync()
                 (wait, emptyResponse) = TryEnsureHeaders();
                 if (wait)
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
+                    using (cancellationToken.Register(s => {
```

Comment:
This cancellation logic seems like it should apply everywhere that GetWaiterTask is called, e.g. ReadDataAsync.

Reply:
In other words, GetWaiterTask should take a CancellationToken, and throw if the CancellationToken is cancelled before the waiter is signalled.

There's some logic to do this with a regular TaskCompletionSource here: https://github.com/dotnet/corefx/blob/af562e84b8b5a4a335ceb5006886a6c42c1c023e/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/TaskCompletionSourceWithCancellation.cs#L10

However, we're not using a regular TaskCompletionSource here. We're using the resettable source. I'm not quite sure the best way to make this work with cancellation.

@stephentoub thoughts?

Reply:
If we need this logic in other places, I agree with @geoffkizer it'd be good to centralize all of this into GetWaiterTask.  The downside to that is it'll make GetWaiterTask more expensive.  But maybe that doesn't matter.

Reply:
@stephentoub Any suggestions re how to make GetWaiterTask cancellable? I'm unclear about the interactions with the resettable task stuff here.

> If we need this logic in other places, I agree with @geoffkizer it'd be good to centralize all of this into GetWaiterTask

We do -- see ReadDataAsync for example, we're ignoring the cancellationToken there currently.

> The downside to that is it'll make GetWaiterTask more expensive. But maybe that doesn't matter.

Only when a cancellationToken is specified, right? (I realize that's the default behavior with HttpClient, but at least it's possible to avoid this if it really matters)

Reply:
When I talk to @stephentoub yesterday, it seems like we always pass cancellationToken token in even if user does not. (To implement timeout and cancelation of all pending requests)